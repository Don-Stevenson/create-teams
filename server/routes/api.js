import express from 'express'
import { body, validationResult, check } from 'express-validator'
import jwt from 'jsonwebtoken'
import bcrypt from 'bcryptjs'
import Player from '../models/Player.js'
import balanceTeams from '../utils/balanceTeams.js'
import validate from '../middleware/validate.js'
import User from '../models/User.js'
import auth from '../middleware/auth.js'
import { upcomingGamesList, rsvpsForGame } from '../utils/getUpcomingGames.js'
import {
  addToBlacklist,
  isBlacklisted,
  getBlacklist,
} from '../utils/tokenBlacklist.js'
import {
  createSession,
  invalidateSession,
  isSessionValid,
  getAllSessions,
} from '../utils/sessionStore.js'

const router = express.Router()

// Public routes (no authentication required)
const publicRouter = express.Router()

// // Registration route
// router.post(
//   '/register',
//   [
//     body('username').trim().isLength({ min: 3 }).escape(),
//     body('password').isLength({ min: 6 }),
//   ],
//   async (req, res) => {
//     const errors = validationResult(req)
//     if (!errors.isEmpty()) {
//       return res.status(400).json({ errors: errors.array() })
//     }

//     try {
//       const user = await User.create(req.body)
//       res.status(201).json({
//         success: true,
//         user: { id: user._id, username: user.username },
//       })
//     } catch (error) {
//       res.status(400).json({ success: false, error: error.message })
//     }
//   }

// Login route
publicRouter.post('/login', async (req, res) => {
  try {
    const { username, password } = req.body
    const user = await User.findOne({ username })
    if (!user) {
      return res.status(401).json({ message: 'Invalid credentials' })
    }
    const isMatch = await bcrypt.compare(password, user.password)
    if (!isMatch) {
      return res.status(401).json({ message: 'Invalid credentials' })
    }
    const token = jwt.sign({ userId: user._id }, process.env.JWT_SECRET, {
      expiresIn: '1h',
    })

    // Create session
    createSession(user._id, token)

    const isProduction = process.env.NODE_ENV === 'production'
    res.cookie('token', token, {
      httpOnly: true,
      secure: isProduction,
      sameSite: isProduction ? 'none' : 'lax', // 'none' for cross-domain in production
      maxAge: 3600000,
      path: '/',
    })
    res.status(200).json({ success: true })
  } catch (error) {
    console.error('Login error:', error)
    res.status(400).json({ success: false, error: error.message })
  }
})

// Test endpoint to force clear all auth state
publicRouter.post('/force-clear-auth', (req, res) => {
  const isProduction = process.env.NODE_ENV === 'production'
  // Clear the token from cookies
  res.clearCookie('token', {
    httpOnly: true,
    secure: isProduction,
    sameSite: isProduction ? 'none' : 'lax',
    path: '/',
  })

  // Force clear all sessions and blacklist (for testing)
  const currentToken = req.cookies?.token
  if (currentToken) {
    addToBlacklist(currentToken)
    invalidateSession(currentToken)
  }

  res.json({ success: true, message: 'All auth state cleared' })
})

// Debug endpoint to see blacklist and session state
publicRouter.get('/debug-auth-state', (req, res) => {
  const token = req.cookies?.token

  if (token) {
    const blacklisted = isBlacklisted(token)
    const sessionValid = isSessionValid(token)

    res.json({
      hasToken: true,
      tokenPreview: token.substring(0, 20) + '...',
      blacklisted,
      sessionValid,
      blacklistSize: getBlacklist().size,
      activeSessionsCount: getAllSessions().length,
    })
  } else {
    res.json({
      hasToken: false,
      blacklistSize: getBlacklist().size,
      activeSessionsCount: getAllSessions().length,
    })
  }
})

// Test endpoint to check auth state
publicRouter.get('/test-auth', (req, res) => {
  if (req.cookies?.token) {
    const blacklisted = isBlacklisted(req.cookies.token)
    const sessionValid = isSessionValid(req.cookies.token)

    try {
      const decoded = jwt.verify(req.cookies.token, process.env.JWT_SECRET)
      res.json({
        hasToken: true,
        blacklisted,
        sessionValid,
        valid: true,
        userId: decoded.userId,
      })
    } catch (error) {
      res.json({
        hasToken: true,
        blacklisted,
        sessionValid,
        valid: false,
        error: error.message,
      })
    }
  } else {
    res.json({ hasToken: false })
  }
})

// Test endpoint to manually clear cookies
publicRouter.get('/clear-cookies', (req, res) => {
  const isProduction = process.env.NODE_ENV === 'production'
  res.clearCookie('token', {
    httpOnly: true,
    secure: isProduction,
    sameSite: isProduction ? 'none' : 'lax',
    path: '/',
  })

  res.status(200).json({ success: true, message: 'Cookies cleared' })
})

// Logout route
publicRouter.post('/logout', (req, res) => {
  // Add current token to blacklist
  const currentToken = req.cookies?.token
  if (currentToken) {
    addToBlacklist(currentToken)
    invalidateSession(currentToken)
  }

  const isProduction = process.env.NODE_ENV === 'production'
  res.clearCookie('token', {
    httpOnly: true,
    secure: isProduction,
    sameSite: isProduction ? 'none' : 'lax',
    path: '/',
  })

  res.status(200).json({ success: true })
})

// Auth check route
publicRouter.get('/auth/check', auth, (req, res) => {
  res
    .status(200)
    .json({ success: true, message: 'Authenticated', userId: req.userId })
})

// Debug endpoint to see cookies and headers
publicRouter.get('/debug-cookies', (req, res) => {
  const cookieHeader = req.headers.cookie
  const allCookies = req.cookies
  const token = req.cookies?.token

  console.log('Debug cookies request:')
  console.log('Cookie header:', cookieHeader)
  console.log('Parsed cookies:', allCookies)
  console.log('Token cookie:', token)

  res.json({
    cookieHeader,
    allCookies,
    hasToken: !!token,
    tokenPreview: token ? token.substring(0, 20) + '...' : null,
    headers: {
      origin: req.headers.origin,
      userAgent: req.headers['user-agent'],
      referer: req.headers.referer,
    },
  })
})

// Protected routes (authentication required)
const protectedRouter = express.Router()
protectedRouter.use(auth)

// Validation rules
const playerValidationRules = [
  body('name').trim().isLength({ min: 2, max: 50 }).escape(),
  body('gameKnowledgeScore').isLength({ min: 1, max: 10 }),
  body('goalScoringScore').isLength({ min: 1, max: 10 }),
  body('attackScore').isInt({ min: 1, max: 10 }),
  body('midfieldScore').isInt({ min: 1, max: 10 }),
  body('defenseScore').isInt({ min: 1, max: 10 }),
  body('fitnessScore').isInt({ min: 1, max: 10 }),
  body('gender').isIn(['male', 'female', 'nonBinary']),
  body('isPlayingThisWeek').isBoolean(),
]

// GET all players
protectedRouter.get('/players', async (req, res, next) => {
  try {
    const players = await Player.find().select('-__v')
    res.json(players)
  } catch (err) {
    next(err)
  }
})

// POST a new player
protectedRouter.post(
  '/players',
  validate(playerValidationRules),
  async (req, res, next) => {
    try {
      const player = new Player(req.body)
      const newPlayer = await player.save()
      res.status(201).json(newPlayer)
    } catch (err) {
      console.error(err)
      next(err)
    }
  }
)

// PUT update player's weekly status
protectedRouter.put(
  '/players/:id',
  validate(playerValidationRules),
  async (req, res, next) => {
    try {
      const updatedPlayer = await Player.findByIdAndUpdate(
        req.params.id,
        { isPlayingThisWeek: req.body.isPlayingThisWeek },
        { new: true, runValidators: true }
      )
      if (!updatedPlayer) {
        return res.status(404).json({ message: 'Player not found' })
      }
      res.json(updatedPlayer)
    } catch (err) {
      next(err)
    }
  }
)

protectedRouter.put(
  '/players/:id/playerInfo',
  validate(playerValidationRules),
  async (req, res, next) => {
    try {
      const {
        name,
        gameKnowledgeScore,
        goalScoringScore,
        attackScore,
        midfieldScore,
        defenseScore,
        fitnessScore,
        isPlayingThisWeek,
        gender,
      } = req.body

      const updatedPlayer = await Player.findByIdAndUpdate(
        req.params.id,
        {
          name,
          gameKnowledgeScore: Number(gameKnowledgeScore),
          goalScoringScore: Number(goalScoringScore),
          attackScore: Number(attackScore),
          midfieldScore: Number(midfieldScore),
          defenseScore: Number(defenseScore),
          fitnessScore: Number(fitnessScore),
          isPlayingThisWeek: Boolean(isPlayingThisWeek),
          gender,
        },
        { new: true, runValidators: true }
      )

      if (!updatedPlayer) {
        return res.status(404).json({ message: 'Player not found' })
      }

      res.json(updatedPlayer)
    } catch (err) {
      console.error('Error updating player:', err)
      if (err.name === 'CastError' || err.name === 'ValidationError') {
        return res
          .status(400)
          .json({ message: 'Invalid data format', error: err.message })
      }
      next(err)
    }
  }
)

// DELETE a player
protectedRouter.delete('/players/:id', async (req, res, next) => {
  try {
    const deletedPlayer = await Player.findByIdAndDelete(req.params.id)
    if (!deletedPlayer) {
      return res.status(404).json({ message: 'Player not found' })
    }
    res.json({ message: 'Player deleted successfully' })
  } catch (err) {
    next(err)
  }
})

// POST balance teams
protectedRouter.post('/balance-teams', express.json(), async (req, res) => {
  try {
    const { numTeams, players } = req.body

    // Validate request body
    if (!req.body) {
      console.error('No request body received')
      return res.status(400).json({ error: 'No request body received' })
    }

    if (!numTeams) {
      console.error('No numTeams provided')
      return res.status(400).json({ error: 'Number of teams is required' })
    }

    const parsedNumTeams = parseInt(numTeams, 10)
    if (isNaN(parsedNumTeams) || parsedNumTeams < 2) {
      console.error('Invalid number of teams:', numTeams)
      return res.status(400).json({ error: 'Invalid number of teams' })
    }

    if (!players) {
      console.error('No players array provided')
      return res.status(400).json({ error: 'Players array is required' })
    }

    // Ensure players is an array
    let playersArray = players
    if (typeof players === 'string') {
      try {
        // Try to parse the string as JSON
        playersArray = JSON.parse(players)
      } catch (e) {
        console.error('Failed to parse players string:', e)
        return res.status(400).json({ error: 'Invalid players data format' })
      }
    }

    if (!Array.isArray(playersArray)) {
      console.error('Players is not an array:', typeof playersArray)
      return res.status(400).json({ error: 'Players must be an array' })
    }

    if (playersArray.length === 0) {
      console.error('Empty players array')
      return res.status(400).json({ error: 'Players array cannot be empty' })
    }

    // Clean up player data
    const cleanedPlayers = playersArray.map((player, index) => {
      if (!player || typeof player !== 'object') {
        console.error(`Invalid player at index ${index}:`, player)
        throw new Error(`Invalid player data structure at index ${index}`)
      }

      const cleanedPlayer = {
        name: String(player.name || ''),
        gameKnowledgeScore: Number(player.gameKnowledgeScore || 0),
        goalScoringScore: Number(player.goalScoringScore || 0),
        attackScore: Number(player.attackScore || 0),
        midfieldScore: Number(player.midfieldScore || 0),
        defenseScore: Number(player.defenseScore || 0),
        fitnessScore: Number(player.fitnessScore || 0),
        gender: String(player.gender || 'male'),
        isPlayingThisWeek: Boolean(player.isPlayingThisWeek),
      }

      // Validate cleaned player data
      if (!cleanedPlayer.name) {
        throw new Error(`Player at index ${index} has no name`)
      }

      if (
        cleanedPlayer.gameKnowledgeScore < 0 ||
        cleanedPlayer.gameKnowledgeScore > 10 ||
        cleanedPlayer.goalScoringScore < 0 ||
        cleanedPlayer.goalScoringScore > 10 ||
        cleanedPlayer.attackScore < 0 ||
        cleanedPlayer.attackScore > 10 ||
        cleanedPlayer.midfieldScore < 0 ||
        cleanedPlayer.midfieldScore > 10 ||
        cleanedPlayer.defenseScore < 0 ||
        cleanedPlayer.defenseScore > 10 ||
        cleanedPlayer.fitnessScore < 0 ||
        cleanedPlayer.fitnessScore > 10
      ) {
        throw new Error(
          `Player ${cleanedPlayer.name} has invalid scores (must be between 0 and 10)`
        )
      }

      if (!['male', 'female', 'nonBinary'].includes(cleanedPlayer.gender)) {
        throw new Error(
          `Player ${cleanedPlayer.name} has invalid gender: ${cleanedPlayer.gender}`
        )
      }

      return cleanedPlayer
    })

    const { teams, totalPlayersPlaying } = balanceTeams(
      cleanedPlayers,
      parsedNumTeams
    )
    return res.json({ teams, totalPlayersPlaying })
  } catch (error) {
    console.error('Error processing request:', error)
    return res.status(400).json({ error: error.message })
  }
})

// Route for bulk updating players is playing this week
protectedRouter.put(
  '/players-bulk-update',
  express.json(),
  async (req, res) => {
    try {
      // Extract and validate the data
      const { isPlayingThisWeek, playerIds } = req.body

      // Validate playerIds
      if (!Array.isArray(playerIds) || playerIds.length === 0) {
        return res.status(400).json({
          success: false,
          error: 'playerIds must be a non-empty array',
        })
      }

      // Ensure all playerIds are strings
      const sanitizedPlayerIds = playerIds.map(id => String(id))

      // Update the players
      const result = await Player.updateMany(
        { _id: { $in: sanitizedPlayerIds } },
        { $set: { isPlayingThisWeek: Boolean(isPlayingThisWeek) } }
      )

      res.json({
        success: true,
        message: 'Players updated successfully',
        modifiedCount: result.modifiedCount,
      })
    } catch (error) {
      console.error('Error updating players:', error)
      res.status(500).json({
        success: false,
        error: 'Failed to update players',
        details: error.message,
      })
    }
  }
)

// GET upcoming games
protectedRouter.get('/upcoming-games', async (req, res) => {
  res.json(upcomingGamesList)
})

// GET rsvps for a game
protectedRouter.get('/rsvps-for-game/:gameId', async (req, res) => {
  const { gameId } = req.params
  try {
    const rsvps = await rsvpsForGame({
      teamId: process.env.TEAM_ID,
      gameId,
    })
    res.json(rsvps)
  } catch (error) {
    console.error('Error fetching RSVPs:', error)
    res.status(500).json({ error: 'Failed to fetch RSVPs' })
  }
})

// Mount the routers
router.use('/', publicRouter)
router.use('/', protectedRouter)

export default router
