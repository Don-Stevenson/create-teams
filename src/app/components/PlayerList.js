import React, { useEffect } from 'react'

const PlayerList = ({
  players,
  onEditPlayer,
  onDeletePlayer,
  fetchPlayers,
}) => {
  useEffect(() => {
    fetchPlayers()
  }, [])
  return (
    <div className="grid grid-cols-[repeat(auto-fill,_minmax(350px,_1fr))] gap-2 w-full p-2">
      {players.map(player => (
        <div key={player._id}>
          <div className="flex justify-between mb-2 p-2 w-80 h-[75px] border border-gray-300 rounded hover:bg-gray-200 ">
            <div className="flex gap-2 p-2 items-center">
              <div>{player.name} </div>
              <div className="text-gray-400 text-xs">
                A:{player.attackScore} D:{player.defenseScore} F:
                {player.fitnessScore}
              </div>
            </div>
            <div className="flex gap-3 items-center">
              <button
                onClick={() => onEditPlayer(player._id)}
                className="border border-gray-300 text-black text-sm font-bold py-1 px-2 rounded h-[45px] w-[60px]"
                data-testid="edit-player"
              >
                Edit
              </button>
              <button
                onClick={() => onDeletePlayer(player._id)}
                data-testid="delete-player"
                className="bg-loonsRed hover:bg-red-900 text-loonsBeige border border-red-900 text-sm font-bold py-1 px-2 h-[45px] rounded w-[60px]"
              >
                Delete
              </button>
            </div>
          </div>
        </div>
      ))}
    </div>
  )
}

export default PlayerList
