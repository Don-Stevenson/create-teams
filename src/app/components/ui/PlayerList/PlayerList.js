import React from 'react'
import PlayerCard from '../Card/PlayerCard'

const PlayerList = ({ players, onEditPlayer, onDeletePlayer }) => {
  const playersByInitial = players.reduce((acc, player) => {
    const initial = player.name.charAt(0).toUpperCase()
    if (!acc[initial]) {
      acc[initial] = []
    }
    acc[initial].push(player)
    return acc
  }, {})

  const sortedInitials = Object.keys(playersByInitial).sort()

  return (
    <div className="flex flex-col gap-4">
      {sortedInitials.map(initial => (
        <div key={initial} className="flex flex-col gap-1">
          <div className="text-2xl font-bold text-slate-700 px-2">
            {initial}
          </div>
          <div className="grid grid-cols-[repeat(auto-fill,_minmax(16rem,_1fr))] xl:grid-cols-[repeat(auto-fill,_minmax(17.5rem,_max-content))] gap-4 w-full p-2">
            {playersByInitial[initial].map(player => (
              <PlayerCard
                key={player._id}
                player={player}
                onEditPlayer={onEditPlayer}
                onDeletePlayer={onDeletePlayer}
              />
            ))}
          </div>
        </div>
      ))}
    </div>
  )
}

export default PlayerList
