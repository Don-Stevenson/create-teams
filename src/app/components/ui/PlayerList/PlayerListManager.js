import PlayerListToggleIsPlaying from './PlayerListToggleIsPlaying'
import { lazy, Suspense } from 'react'

const LoadingPlayers = lazy(() => import('../Loading/LoadingPlayers'))

export default function PlayerListManager({
  players,
  selectedPlayerCount,
  selectAll,
  openPlayerList,
  showLoadingMessage,
  queryRsvpsForGame,
  bulkUpdateMutation,
  onTogglePlayingThisWeek,
  onSelectAll,
  onTogglePlayerList,
}) {
  return (
    <div className="print:hidden">
      <div className="flex-col flex-wrap">
        <h2 className="text-3xl font-semibold mb-4 print:hidden md:justify-center text-loonsDarkBrown">
          Player List
        </h2>
        <p className="flex text-center md:justify-center mb-4 print:hidden">
          <span className="font-bold text-xl text-gray-800 ">
            {`Total Players Selected: ${selectedPlayerCount}`}
          </span>
        </p>
        <div className="flex justify-center items-center mb-2">
          <button
            onClick={onTogglePlayerList}
            className="bg-loonsRed hover:bg-red-900 text-loonsBeige border-red-900 mb-2 text-lg font-bold border-2  rounded-md px-2 py-1"
          >
            {openPlayerList ? 'Hide Player List' : 'Show Player List'}
          </button>
        </div>
        {openPlayerList && (
          <>
            <div className="flex md:justify-center mb-4 print:hidden">
              <label className="inline-flex items-center">
                <input
                  type="checkbox"
                  className="form-checkbox h-5 w-5 text-loonsRed"
                  checked={selectAll}
                  onChange={onSelectAll}
                  disabled={bulkUpdateMutation.isPending}
                />
                <span className="ml-2 text-gray-700 text-sm">
                  Toggle All Players Playing / Not Playing
                </span>
              </label>
            </div>
            {players.length !== 0 && (
              <Suspense fallback={<LoadingPlayers />}>
                <PlayerListToggleIsPlaying
                  players={players}
                  rsvpsForGame={queryRsvpsForGame}
                  onTogglePlayingThisWeek={onTogglePlayingThisWeek}
                />
              </Suspense>
            )}
          </>
        )}
      </div>
    </div>
  )
}
