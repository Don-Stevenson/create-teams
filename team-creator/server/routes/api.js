// server/routes/api.js
import express from 'express'
const router = express.Router()
import { check } from 'express-validator'
import Player from '../models/Player.js'
import balanceTeams from '../utils/teamBalancer.js'
import validate from '../middleware/validate.js'

// Validation rules
const playerValidationRules = [
  check('name').trim().isLength({ min: 2, max: 50 }).escape(),
  check('attackScore').isInt({ min: 0, max: 100 }),
  check('defenseScore').isInt({ min: 0, max: 100 }),
  check('fitnessScore').isInt({ min: 0, max: 100 }),
  check('gender').isIn(['male', 'female', 'other']),
]

// GET all players
router.get('/players', async (req, res, next) => {
  try {
    const players = await Player.find().select('-__v')
    res.json(players)
  } catch (err) {
    next(err)
  }
})

// POST a new player
router.post(
  '/players',
  validate(playerValidationRules),
  async (req, res, next) => {
    try {
      const player = new Player(req.body)
      const newPlayer = await player.save()
      res.status(201).json(newPlayer)
    } catch (err) {
      next(err)
    }
  }
)

// PUT update player's weekly status
router.put(
  '/players/:id',
  [check('isPlayingThisWeek').isBoolean()],
  validate([check('isPlayingThisWeek').isBoolean()]),
  async (req, res, next) => {
    try {
      const updatedPlayer = await Player.findByIdAndUpdate(
        req.params.id,
        { isPlayingThisWeek: req.body.isPlayingThisWeek },
        { new: true, runValidators: true }
      )
      if (!updatedPlayer) {
        return res.status(404).json({ message: 'Player not found' })
      }
      res.json(updatedPlayer)
    } catch (err) {
      next(err)
    }
  }
)

// POST balance teams
router.post(
  '/balance-teams',
  validate([check('numTeams').isInt({ min: 2, max: 10 })]),
  async (req, res, next) => {
    try {
      const playingPlayers = await Player.find({ isPlayingThisWeek: true })
      const balancedTeams = balanceTeams(playingPlayers, req.body.numTeams)
      res.json(balancedTeams)
    } catch (err) {
      next(err)
    }
  }
)

export default router
