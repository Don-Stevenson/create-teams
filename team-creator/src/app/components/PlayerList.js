// components/PlayerList.js
export default function PlayerList({ players }) {
  return (
    <div className="bg-white shadow-md rounded px-8 pt-6 pb-8 mb-4">
      <ul className="divide-y divide-gray-200">
        {players.map(player => (
          <li key={player._id} className="py-4">
            <div className="flex justify-between">
              <span className="font-semibold">{player.name}</span>
              <span className="text-gray-500">{player.gender}</span>
            </div>
            <div className="mt-2 text-sm text-gray-500">
              Attack: {player.attackScore} | Defense: {player.defenseScore} |
              Fitness: {player.fitnessScore}
            </div>
          </li>
        ))}
      </ul>
    </div>
  )
}
