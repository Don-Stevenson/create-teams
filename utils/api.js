// utils/api.js
import axios from 'axios'
import config_url from '../config'

const api = axios.create({
  baseURL: `${config_url}/api/`,
  headers: {
    'Content-Type': 'application/json',
  },
})

api.interceptors.response.use(
  response => response,
  error => {
    console.error('API Error:', error.response)
    return Promise.reject(error)
  }
)

export default api
