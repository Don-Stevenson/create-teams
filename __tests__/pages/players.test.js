import { render, screen, fireEvent, waitFor, act } from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import Players from '../../pages/players'
import api from '../../utils/api'
import PlayerList from '../../src/app/components/PlayerList'

jest.mock('../../utils/api')
jest.mock('../../src/app/components/withAuth', () => {
  return jest.fn(Component => {
    const WithAuthComponent = props => <Component {...props} />
    WithAuthComponent.displayName = `withAuth(${Component.name})`
    return WithAuthComponent
  })
})
jest.mock('../../src/app/components/Layout', () => {
  return jest.fn(({ children }) => (
    <div data-testid="mock-layout">{children}</div>
  ))
})

const mockPlayers = [
  {
    _id: '1',
    name: 'John Doe',
    attackScore: 6,
    defenseScore: 7,
    fitnessScore: 8,
    isPlayingThisWeek: true,
  },
  {
    _id: '2',
    name: 'Alice Smith',
    attackScore: 7,
    defenseScore: 8,
    fitnessScore: 9,
    isPlayingThisWeek: false,
  },
]

describe('Players Page', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    api.get.mockResolvedValue({ data: mockPlayers })
  })

  describe('Rendering', () => {
    it('renders the page with correct headings', async () => {
      await act(async () => {
        render(<Players />)
      })

      expect(screen.getByText('Manage Players')).toBeInTheDocument()
      expect(screen.getByText('Add A New Player')).toBeInTheDocument()
      expect(screen.getByText('List of Players')).toBeInTheDocument()
    })

    it('fetches and displays players on mount', async () => {
      await act(async () => {
        render(<Players />)
      })

      await waitFor(() => {
        expect(api.get).toHaveBeenCalledWith('/players')
        expect(screen.getByText('John Doe')).toBeInTheDocument()
        expect(screen.getByText('Alice Smith')).toBeInTheDocument()
      })
    })
  })

  describe('Add Player Functionality', () => {
    let user

    beforeEach(() => {
      user = userEvent.setup()
      jest.clearAllMocks()
    })

    it('toggles add player form visibility', async () => {
      await act(async () => {
        render(<Players />)
      })
      const addButton = screen.getByText('Add A New Player')
      await user.click(addButton)
      expect(screen.getByTestId('add-player-form')).toBeInTheDocument()
      await user.click(screen.getByText('cancel'))
      expect(screen.queryByTestId('add-player-form')).not.toBeInTheDocument()
    })

    it('adds a new player successfully', async () => {
      const newPlayer = {
        _id: '3',
        name: 'New Player',
        gameKnowledgeScore: '2',
        goalScoringScore: '2',
        attackScore: '2',
        midfieldScore: '2',
        defenseScore: '2',
        fitnessScore: '2',
        gender: 'male',
        isPlayingThisWeek: true,
      }

      api.post.mockResolvedValueOnce({ data: newPlayer })

      await act(async () => {
        render(<Players />)
      })

      const addButton = screen.getByText('Add A New Player')
      await act(async () => {
        await user.click(addButton)
      })

      await act(async () => {
        await user.type(screen.getByLabelText(/name/i), newPlayer.name)
        await user.type(
          screen.getByLabelText(/game knowledge score/i),
          newPlayer.gameKnowledgeScore
        )
        await user.type(
          screen.getByLabelText(/goal scoring score/i),
          newPlayer.goalScoringScore
        )
        await user.type(
          screen.getByLabelText(/attack score/i),
          newPlayer.attackScore
        )
        await user.type(
          screen.getByLabelText(/midfield score/i),
          newPlayer.midfieldScore
        )
        await user.type(
          screen.getByLabelText(/defense score/i),
          newPlayer.defenseScore
        )
        await user.type(
          screen.getByLabelText(/Mobility\/Stamina/i),
          newPlayer.fitnessScore
        )
        screen.getByTestId('gender')
      })

      await act(async () => {
        await user.click(screen.getByRole('button', { name: /add player/i }))
      })

      await waitFor(() => {
        expect(api.post).toHaveBeenCalledWith(
          '/players',
          expect.objectContaining({
            name: newPlayer.name,
            goalScoringScore: newPlayer.goalScoringScore,
            gameKnowledgeScore: newPlayer.gameKnowledgeScore,
            attackScore: newPlayer.attackScore,
            midfieldScore: newPlayer.midfieldScore,
            defenseScore: newPlayer.defenseScore,
            fitnessScore: newPlayer.fitnessScore,
            isPlayingThisWeek: newPlayer.isPlayingThisWeek,
          })
        )
      })
    })

    describe('Edit Player Functionality', () => {
      it('opens edit modal when edit button is clicked', async () => {
        await act(async () => {
          render(<Players />)
        })

        await waitFor(() => {
          const editButton = screen.getAllByTestId('edit-player')[0]
          fireEvent.click(editButton)
        })
        expect(screen.getByTestId('edit-player-modal')).toBeInTheDocument()
      })

      describe('PlayerList', () => {
        it('calls onEditPlayer when edit button is clicked', () => {
          const mockOnEditPlayer = jest.fn()
          render(
            <PlayerList
              players={mockPlayers}
              onEditPlayer={mockOnEditPlayer}
              onDeletePlayer={() => {}}
              fetchPlayers={() => {}}
            />
          )

          const editButton = screen.getAllByTestId('edit-player')[0]
          fireEvent.click(editButton)

          expect(mockOnEditPlayer).toHaveBeenCalledWith('1')
        })
      })
    })

    describe('Delete Player Functionality', () => {
      it('shows delete confirmation modal', async () => {
        await act(async () => {
          render(<Players />)
        })

        await waitFor(() => {
          const deleteButton = screen.getAllByTestId('delete-player')[1]
          fireEvent.click(deleteButton)
        })
        expect(
          screen.getByText(/Are you sure you want to delete John Doe?/i)
        ).toBeInTheDocument()
      })

      it('deletes player when confirmed', async () => {
        api.delete.mockResolvedValue({})

        await act(async () => {
          render(<Players />)
        })

        await waitFor(async () => {
          const deleteButton = screen.getAllByTestId('delete-player')[1]
          fireEvent.click(deleteButton)

          const confirmDelete = screen.getAllByText('Delete')[2]
          fireEvent.click(confirmDelete)
        })
        expect(api.delete).toHaveBeenCalledWith('/players/1')
      })

      it('cancels deletion when cancel is clicked', async () => {
        await act(async () => {
          render(<Players />)
        })

        await waitFor(() => {
          const deleteButton = screen.getAllByTestId('delete-player')[0]
          fireEvent.click(deleteButton)

          const cancelButton = screen.getByText('Cancel')
          fireEvent.click(cancelButton)
        })

        expect(
          screen.queryByText(/Are you sure you want to delete/)
        ).not.toBeInTheDocument()
        expect(screen.getByText('John Doe')).toBeInTheDocument()
      })
    })

    describe('Error Handling', () => {
      it('handles fetch error gracefully', async () => {
        const consoleSpy = jest
          .spyOn(console, 'error')
          .mockImplementation(() => {})
        api.get.mockRejectedValue(new Error('Failed to fetch'))

        await act(async () => {
          render(<Players />)
        })

        await waitFor(() => {
          expect(consoleSpy).toHaveBeenCalledWith(
            'Failed to fetch players:',
            expect.any(Error)
          )
        })

        consoleSpy.mockRestore()
      })

      it('handles update error gracefully', async () => {
        const consoleSpy = jest.spyOn(console, 'error').mockImplementation()

        const mockPlayers = [
          {
            _id: '1',
            name: 'John Doe',
            attackScore: 2,
            defenseScore: 2,
            fitnessScore: 2,
            isPlayingThisWeek: true,
          },
        ]

        api.get.mockResolvedValue({ data: mockPlayers })
        const mockError = new Error('Update failed')
        api.put.mockRejectedValue(mockError)

        render(<Players />)

        await screen.findByText('John Doe')
        const editButton = screen.getByTestId('edit-player')
        fireEvent.click(editButton)

        const modal = await screen.findByTestId('edit-player-modal')
        expect(modal).toBeInTheDocument()

        const nameInput = screen.getByLabelText('Name')
        await act(async () => {
          fireEvent.change(nameInput, { target: { value: 'John Doe Updated' } })
        })

        const saveChangesButton = screen.getByText(/Save Changes/i)
        await act(async () => {
          fireEvent.submit(saveChangesButton)
        })

        expect(console.error).toHaveBeenCalledWith(
          'Failed to update player:',
          mockError
        )

        expect(modal).toBeInTheDocument()
        expect(nameInput.value).toBe('John Doe Updated')

        consoleSpy.mockRestore()
      })
    })

    it('matches snapshot', () => {
      const { container } = render(<Players />)
      expect(container).toMatchSnapshot()
    })
  })
})
