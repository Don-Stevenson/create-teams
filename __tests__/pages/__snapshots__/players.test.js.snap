// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Players Page Add Player Functionality matches snapshot 1`] = `
<div>
  <div
    data-testid="mock-layout"
  >
    <div
      class="flex-col px-4 py-8"
    >
      <h1
        class="text-3xl font-bold mb-8 text-loonsDarkBrown"
      >
        Manage Players
      </h1>
      <div>
        <button
          class="text-lg font-semibold mb-4 rounded bg-loonsRed hover:bg-red-900 text-loonsBeige border border-red-900 w-[200px] h-18 p-3 text-center"
        >
          Add A New Player
        </button>
      </div>
      <div>
        <h2
          class="text-2xl font-semibold mb-4 text-black"
        >
          List of Players
        </h2>
        <div
          class="flex-col text-black mb-4 gap-1"
        >
          <p
            class="text-sm"
          >
            Legend:
          </p>
          <p
            class="text-xxs"
          >
            K: Game knowledge, Sc: Goal Scoring, A: Attack, Md: Midfield, D: Defense, M/S: Mobility/Stamina
          </p>
        </div>
        <div
          class="text-center py-4"
        >
          Loading players...
        </div>
      </div>
    </div>
  </div>
</div>
`;
